include/swift/AST/EvaluatorDependencies.h

```diff
@@ -0,0 +1,88 @@
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file defines data structures to support the request evaluator's
+// automatic incremental dependency tracking functionality.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_AST_EVALUATOR_DEPENDENCIES_H
+#define SWIFT_AST_EVALUATOR_DEPENDENCIES_H
+
+#include "swift/AST/AttrKind.h"
+#include "swift/AST/SourceFile.h"
+#include "llvm/ADT/PointerIntPair.h"
+
+namespace swift {
+
+namespace evaluator {
+
+/// The "scope" of a dependency edge tracked by the evaluator.
+///
+/// Dependency scopes come in two flavors: cascading and private. A private
+/// edge captures dependencies discovered in contexts that are not visible to
+/// to other files. For example, a conformance to a private protocol, or the use
+/// of any names inside of a function body. A cascading edge, by contrast,
+/// captures dependencies discovered in the remaining visible contexts. These
+/// are types with at least \c internal visibility, names defined or used
+/// outside of function bodies with at least \c internal visibility, etc. A
+/// dependency that has cascading scope is so-named because upon traversing the
+/// edge, a reader such as the driver should continue transitively evaluating
+/// further dependency edges.
+///
+/// A cascading edge is always conservatively correct to produce, but it comes
+/// at the cost of increased resources spent (and possibly even wasted!) during
+/// incremental compilation. A private edge, by contrast, is more efficient for
+/// incremental compilation but it is harder to safely use.
+///
+/// To ensure that these edges are registered consistently with the correct
+/// scopes, requests that act as the source of dependency edges are required
+/// to specify a \c DependencyScope under which all evaluated sub-requests will
+/// register their dependency edges. In this way, \c DependencyScope values
+/// form a stack-like structure and are pushed and popped by the evaluator
+/// during the course of request evaluation.
+///
+/// When determining the kind of scope a request should use, always err on the
+/// side of a cascading scope unless there is absolute proof any discovered
+/// dependencies will be private. Inner requests may also defensively choose to
+/// flip the dependency scope from private to cascading in the name of safety.
+enum class DependencyScope : bool {
+  Private = false,
+  Cascading = true,
+};
+
+/// Returns a \c DependencyScope appropriate for the given (formal) access level.
+inline DependencyScope getScopeForAccessLevel(AccessLevel l) {
+  switch (l) {
+  case AccessLevel::Private:
+  case AccessLevel::FilePrivate:
+    return DependencyScope::Private;
+  case AccessLevel::Internal:
+  case AccessLevel::Public:
+  case AccessLevel::Open:
+    return DependencyScope::Cascading;
+  }
+}
```

Comment:
How does this jibe with the inside-function-body part of dependency scoping? Is that determination somewhere else?

Reply:
Right now, it doesn't. It is used in 4 places and all of them are the exceptions I noted in the conversation above. We should remove this function once we transition to a more consistent ontology of dependency management.